src/shared/vpick.c

```diff
@@ -682,6 +682,22 @@ int path_pick_update_warn(
         return 1;
 }
 
+int path_pick_raw_or_dir(
+                const char *toplevel_path,
+                int toplevel_fd,
+                const char *path,
+                PickFlags flags,
+                PickResult *ret) {
+
+        int r;
+
+        r = path_pick(toplevel_path, toplevel_fd, path, &pick_filter_image_raw, flags, ret);
+        if (r == -EISDIR)
+                r = path_pick(toplevel_path, toplevel_fd, path, &pick_filter_image_dir, flags, ret);
+
+        return r;
+}
```

Comment:
i find this approach a bit hackish tbh, since it operates with two filters that are applied entirely independently, enumerating the entries twice.

I think a more correct approach would be to have a single PickFilter definition covering both.

i.e. `const PickFilter pick_filter_image_any`. This would be mostly trivial to implement, except for one things: we'd have to change the `.suffix` field to take list of suffixes (so that we can list both ".raw" first, and "" second). Thus we'd have:

```c
const PickFilter pick_filter_image_any {
        .type_mask = (UINT32_C(1) << DT_REG) | (UINT32_C(1) << DT_BLK),
        .architecture = _ARCHITECTURE_INVALID,
        .suffix = STRV_MAKE(".raw", ""),
};
```

This would then require that in places where we currently do `endswith(…, filter->suffix)` we instead to use endswith_strv()`...

(And it would also mean that in format_fname() if more than 0 or 1 suffixes are specified we return -ENOXEC, since the filter alone would not pin a filename, it leaves ambiguity.)

Reply:
doing this is kinda necessary anyway, since even for the existing pick_filter_image_raw sruct we should actually list .sysext.raw and .confext.raw before .raw as suffix